### PR TITLE
Memory Pool strawman

### PR DESCRIPTION
**Pull request by rynowak**
This is a very bare-bones implementation so we can integrate it in a few places and start iterating on our whether it meets our needs.

### CODE REVIEW

File: test/Microsoft.Framework.MemoryPool.Test/project.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        "Microsoft.Framework.MemoryPool": "",
+        "xunit.runner.aspnet": "2.0.0-aspnet-*"
+    },
+    "commands": {
+        "test": "xunit.runner.aspnet"
+    },
+    "frameworks": {
+        "dnx451": {
+            "dependencies": {
+                "Moq": "4.2.1312.1622"
```
Comment (davidfowl) [inline test/Microsoft.Framework.MemoryPool.Test/project.json]: Avoid using Moq where possible until we have one for CoreCLR

File: test/Microsoft.Framework.MemoryPool.Test/project.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿{
+    "compilationOptions": {
+        "warningsAsErrors": true
+    },
+    "dependencies": {
+        "Microsoft.Framework.MemoryPool": "",
```
Comment (davidfowl) [inline test/Microsoft.Framework.MemoryPool.Test/project.json]: We use 1.0.0-* even for project references.

**Comment by rynowak** [comment]
Updated

File: src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            _isDisposed = true; // Stops anything from being returned to the pool.
+
+            DefaultLeasedArraySegment segment;
+            while (_segments.TryDequeue(out segment))
+            {
+                segment.Destroy();
+            }
+        }
+
+        private class DefaultLeasedArraySegment : LeasedArraySegment<T>
```
Comment (davidfowl) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: Did you want to preserve separation between the poco and the array segment leased by a particular pool implementation? Why not put tis logic in the default impl

Comment (rynowak) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: You need to know something about the pool to implement the finalizer. Likewise this avoids putting `Destroy` on the API the user sees

Comment (rynowak) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: If you want we could simplify all of this for now by just skipping the finalizer. I wanted to include it in the POC

Comment (davidfowl) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: What bugs me is that we downcast expect it to be a DefaultLeasedArraySegment

Comment (rynowak) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: > Likewise this avoids putting Destroy on the API the user sees

Do you care about this? Alternative here would be to explicitly implement an interface. 

We wouldn't support returning to a pool other than `Owner`, so the downcast is safe.

Comment (davidfowl) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: Ok I'd be happy with that check in Return.

File: src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            {
+                segment = new DefaultLeasedArraySegment(new ArraySegment<T>(new T[BlockSize]), this);
+            }
+
+            return segment;
+        }
+
+        public void Return(LeasedArraySegment<T> buffer)
+        {
+            var segment = (DefaultLeasedArraySegment)buffer;
```
Comment (davidfowl) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: Check `buffer.Owner == this` and throw if it isn't

File: src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+{
+    public class DefaultArraySegmentPool<T> : IArraySegmentPool<T>, IDisposable
+    {
+        public readonly static int Capacity = Environment.ProcessorCount * 4;
+        public readonly static int BlockSize = 4096;
+
+        private readonly ConcurrentQueue<DefaultLeasedArraySegment> _segments;
+        private volatile bool _isDisposed;
+
+        public DefaultArraySegmentPool()
```
Comment (davidfowl) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: nuke this and new it up inline.

Comment (rynowak) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: Constructor > 150 character long line

Comment (davidfowl) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: Hit enter

Comment (rynowak) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: Enter key is broken.

I use powershell to insert `Environment.NewLine` everywhere I need it.

**Comment by rynowak** [comment]
@davidfowl any other feedback?

**Comment by rynowak** [comment]
updated

File: src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+namespace Microsoft.Framework.MemoryPool
+{
+    public class DefaultArraySegmentPool<T> : IArraySegmentPool<T>, IDisposable
+    {
+        public readonly static int Capacity = Environment.ProcessorCount * 4;
+        public readonly static int BlockSize = 4096;
+
+        private readonly ConcurrentQueue<DefaultLeasedArraySegment> _segments = 
+            new ConcurrentQueue<DefaultLeasedArraySegment>();
```
Comment (rynowak) [inline src/Microsoft.Framework.MemoryPool/DefaultArraySegmentPool.cs]: See now it's super ugly like you wanted

**Pull request closed without merging**